### ISSUE TITLE
Change color matching from RGB to HSV

### ISSUE THREAD

**Issue by DavidB-CMU** [personalrobotics/block_sorting#16]
@aaronjoh suggested changing from RGB to HSV color space when matching detected block colors, to make it more robust.

**Comment by DavidB-CMU** [comment]
HSV color space is still non-linear for the "hue" component.
HERB is now matching colors in the Lab color space which is linear in the amount of red-green-blue-yellow.

**Comment by siddhss5** [comment]
This was still a big issue at the demo. What's the update on it @DavidB-CMU?

**Comment by DavidB-CMU** [comment]
@siddhss5 I can collect some data on what colors HERB is seeing under various illumination conditions

**Comment by siddhss5** [comment]
That's an excellent idea and a very useful dataset to have. Let's do it!

**Comment by mklingen** [comment]
@DavidB-CMU where is the code that does Lab space color segmentation?

**Comment by DavidB-CMU** [comment]
@mklingen In "convert_color.py" of the "issue/16" branch of the "block_sorting" package:
https://github.com/personalrobotics/block_sorting/blob/feature/issue16/src/block_sorting/convert_color.py

getClosestColorLab()
Compares the euclidean distance between the a,b components of a known color with a dict of four colors.

**Comment by DavidB-CMU** [comment]
@siddhss5 So far I cannot reproduce the problem in the lab.

I made the lighting as dark as possible and HERB can still distinguish the 4 block colors in the RGB space!
Apparently the museum had unusual light bulbs so maybe they caused a color shift.

**Comment by mklingen** [comment]
You want to change the hue of the lighting to break the kinect's white balance. If we get a colored lamp (or put colored paper in front of a strong lamp) we should be able to break it pretty easily. I heard that the demo was done under sunlight. Its possible the sensor was *over* exposed.

Ultimately to make it as robust as possible we should do two things:

* Some kind of white balance calibration for the kinect
* A real classifier (SVM, etc.) trained from real data.

**Comment by mklingen** [comment]
For white balance, we can put a white piece of paper in the scene under the lighting conditions of the demo, then run something like this: https://gist.github.com/tomykaira/94472e9f4921ec2cf582 on the color image before getting the blocks.

**Comment by Shushman** [comment]
I inspected the colour config file after the demo hours. A primary reason for the error was that in haste, I think you specified 2 yellows - as "yellow1" and "yellow2", and there was no "blue" - and so there was no chance of a blue being detected and the "yellow2" was closer to the green than the stated "green". Once the ambient light came, the actual RGB values that the blocks would give were actually reasonable.

So on Friday evening I changed the config file to reflect a more reasonable colour distribution, and so it worked decently on Saturday with ambient lighting and that's probably why it works in the lab too.

**Comment by DavidB-CMU** [comment]
Yes if either of the RGB or Lab matching functions were missing a known color then they could not match it  :-)

**Comment by DavidB-CMU** [comment]
I had HERB look at some blocks on the table in the lab and collected 18 samples for each block, with varying ambient light.

Here are the RGB colors as detected by the Kinect:
![block_sorting_rgb](https://cloud.githubusercontent.com/assets/14189705/10109925/198714ea-6398-11e5-848c-b24477f419f3.png)

The darker spots for each color represent the samples collected in darker ambient light.
You can't tell just by looking at the above picture, but the interesting thing about the 4 clusters is that each cluster of points is on a 2D plane.

Here are the same samples converted in the Lab color space:
![block_sorting_lab](https://cloud.githubusercontent.com/assets/14189705/10110064/db0314d4-6398-11e5-86c8-9b6b88bcd162.png)

Using the euclidean distance, the samples which are closest together appear to be closer together in the Lab space than in the RGB space.
